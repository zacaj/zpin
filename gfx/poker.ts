import { Group, ImageView } from "aminogfx-gl";
import { Poker, Card, getFileForCard } from "../poker";
import { gfx, makeImage, Screen, Image } from "../gfx";
import { onChange } from "../state";

export class PokerGfx extends Group {
    constructor(
        public poker: Poker,
    ) {
        super(gfx);

        this.add(new PokerHand(poker, poker.player).x(Screen.w/2).y(Screen.h/2-PokerHand.h*1.1));
        this.add(new PokerHand(poker, poker.dealer).x(Screen.w/2).y(Screen.h/2+PokerHand.h*1.1));
    }
}

class PokerHand extends Group {
    static w = 800;
    static h = 150;
    constructor(
        public poker: Poker,
        public hand: (Card|null)[],
    ) {
        super(gfx);
        this.originX(0.5).originY(0.5);
        for (let i=0; i<7; i++) {
            const img = makeImage('', 100, PokerHand.h, false);
            img.x(PokerHand.w/7*this.children.length-PokerHand.w/2);
            this.add(img);
        }

        poker.listen(onChange(hand), () => this.refresh());
        this.refresh();
    }

    refresh() {
        let i=0;
        for (const card of this.hand) {
            const file = getFileForCard(card);
            Image.set(this.children[i++] as ImageView, file);
        }
    }
}