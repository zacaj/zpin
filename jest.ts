import { resetMachine, MachineOutputs, machine, Solenoid } from './machine';
import { resetSwitchMatrix } from './switch-matrix';
import { Events } from './events';
import { MPU } from './mpu';
import { Timer, setTime, wait } from './timer';
import { Log } from './log';

beforeEach(async () => {
    jest.spyOn(Timer, 'schedule').mockRestore();
    jest.spyOn(Events, 'listen').mockRestore();
    jest.spyOn(Events, 'fire').mockRestore();
    jest.spyOn(Timer, 'callIn').mockRestore();
    jest.spyOn(Log, 'init').mockImplementation(() => { throw 'unexpected' });
    jest.spyOn(Log, 'write').mockReturnValue();
    jest.spyOn(Log, 'logMessage').mockReturnValue();
    jest.spyOn(Log, 'trace').mockReturnValue();
    Timer.reset();
    Events.resetAll();
    resetSwitchMatrix();
    resetMachine();
    await setTime(1);
    jest.spyOn(MPU, 'sendCommandCode').mockImplementation(async (cmd) => {
        debugger;
        
        expect(cmd).toBe('mocked');

        return {
            code: 200,
            resp: 'mocked',
        };
    });
});

afterEach(async () => {
    Events.resetAll();
    Timer.reset();
    await setTime();
    jest.spyOn(Timer, 'callIn').mockReturnValue({} as any);
    jest.spyOn(Timer, 'schedule').mockReturnValue({} as any);
    jest.spyOn(Events, 'listen').mockReturnValue({} as any);
    jest.spyOn(Events, 'fire').mockReturnValue({} as any);
    await new Promise(r => setTimeout(r, 50));
});