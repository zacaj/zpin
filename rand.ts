import seedrandom = require('seedrandom');

export class Rng {

    private rng!: seedrandom.prng;
    constructor(
        seed = Math.random().toString(),
    ) {
        this.rng = seedrandom(seed);
    }

    rand(): number {
        return this.rng();
    }

    randRange(start: number, end: number): number {
        return Math.floor(this.rng()*(end-start+1)+start);
    }

    weightedRand(...weights: number[]): number {
        let sum = weights.reduce((prev, cur) => prev+cur, 0);
        const rand = Math.floor(this.rng()*sum);
        for (let i=0; i<weights.length; i++) {
            sum -= weights[i];
            if (sum <= rand) return i;
        }
        return weights.length - 1;
    }

    weightedSelect<T>(...weights: [number, T][]): T {
        return weights[this.weightedRand(...weights.map(x => x[0]))][1];
    }

    weightedRange(...weights: [number, number, number?][]): number {
        const i = this.weightedRand(...weights.map(x => x[0]));
        const weight = weights[i];
        if (weight[2])
            return this.randRange(weight[1], weight[2]);
        else 
            return weight[1];
    }

    randSelect<T>(...values: T[]): T {
        return this.randSelectMany(1, ...values)[0];
    }
    randSelectRange<T>(start: number, end: number, ...values: T[]): T[] {
        return this.randSelectMany(this.randRange(start, end), ...values);
    }
    randSelectMany<T>(count: number, ...values: T[]): T[] {
        const ret: T[] = [];
        while (ret.length < count) {
            const v = values[this.randRange(0, values.length-1)];
            if (!ret.includes(v))
                ret.push(v);
        }
        return ret;
    }

    shuffle<T>(arr: T[], times = 3): T[] {
        for (let k=0; k<times; k++) {
            for (let i = arr.length - 1; i > 0; i--) {
                const j = Math.floor(this.rand() * (i + 1));
                [arr[i], arr[j]] = [arr[j], arr[i]];
            }
        }
        return arr;
    }
}