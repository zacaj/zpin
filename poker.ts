import { MachineOutputs, machine } from './machine';
import { Mode } from './mode';
import { State } from './state';
import { Outputs } from './outputs';
import { DropDownEvent } from './drop-bank';
import { Group } from 'aminogfx-gl';
import { gfx, screen } from './gfx';
import { PokerGfx } from './gfx/poker';

export class Poker extends Mode<MachineOutputs> {
    player: (Card|null)[] = [];
    dealer: (Card|null)[] = [];

    deck = makeDeck();

    step = 2;

    slots: (Card|null)[] = [];

    gfx!: PokerGfx;

    constructor() {
        super();
        State.declare<Poker>(this, ['player', 'dealer', 'slots']);
        this.deal();

        const outs: any  = {};
        for (const target of machine.dropTargets) {
            outs[target.image.name] = () => getFile(this.slots[target.num]);
        }
        this.out = new Outputs(this, outs);

        this.listen(e => e instanceof DropDownEvent, (e: DropDownEvent) => {
            const target = e.target;
            if (this.slots[target.num]) {
                this.player[this.step] = this.slots[target.num];
                this.slots[target.num] = null;
                this.dealer[this.step] = this.deck.pop()!;
                this.step++;

                if (this.step >= 7) {
                    this.deal();
                }
            }
        });

        screen.add(this.gfx = new PokerGfx(this));
    }

    end() {
        screen.remove(this.gfx);
        return super.end();
    }

    deal() {
        this.player.clear();
        this.dealer.clear();
        this.slots.clear();
        this.deck = makeDeck();

        this.player.push(this.deck.pop()!);
        this.player.push(this.deck.pop()!);
        this.dealer.push(this.deck.pop()!);
        this.dealer.push(this.deck.pop()!);
        this.dealer[0]!.flipped = true;
        this.dealer[1]!.flipped = true;
        for (let i=0; i<5; i++) {
            this.player.push(null);
            this.dealer.push(null);
        }

        for (let i=0; i<20; i++) {
            this.slots.push(this.deck.pop()!);
        }

        this.step = 2;
    }
}

function getFile(card: Card|null) {
    if (!card) return 'empty';
    if (card.flipped) return 'back';
    // return card.num+card.suit+'-s';
    let num = `${card.num}`;
    if (card.num>10) num = 'JQK'.charAt(card.num-11);
    if (num === '1') num = 'A';
    return num+card.suit.toUpperCase();
}
export const getFileForCard = getFile;

function makeDeck(): Card[] {
    const deck: Card[] = [];
    for (let i=1; i<=13; i++) {
        for (const suit of Object.values(Suit)) {
            deck.push({num: i, suit});
        }
    }
    for (let i = deck.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [deck[i], deck[j]] = [deck[j], deck[i]];
    }
    return deck;
}

export interface Card {
    num: number;
    suit: Suit; 
    flipped?: true;
}

enum Suit {
    Hearts = 'h',
    Spades = 's',
    Clubs = 'c',
    Diamonds = 'd',
}