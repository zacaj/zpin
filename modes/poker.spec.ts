import { Suit, Card, findPairs, findFlushes, findStraights, bestHand, compareHands, Poker } from './poker';
import { num } from '../util';
import { machine } from '../machine';
import { playRecording, continueRecording, testRecording, finishRecording } from '../recording';
import { passTime } from '../timer';
import { settleForks } from '../promises';
import { snapshot } from '../jest';
import { Rng } from '../rand';

describe('poker', () => {
    function hand(str: string): Card[] {
        return str.split(',').map(s => ({
            suit: s.charAt(s.length===3? 2:1) as Suit,
            num: num(s.slice(0, s.length-1)),
        }));
    }
    test('findPairs', () => {
        expect(findPairs(hand('2h,2d,3s'))).toEqual([hand('2h,2d')]);
        expect(findPairs(hand('2h,2d,3s,5s,1d'))).toEqual([hand('2h,2d')]);
        expect(findPairs(hand('2h,2d,3s,3d'))).toEqual([hand('3s,3d'), hand('2h,2d')]);
        expect(findPairs(hand('2h,2d,2s,2c'))).toEqual([hand('2h,2d,2s,2c')]);
        expect(findPairs(hand('2h,2d,2s,3s,3d'))).toEqual([hand('2h,2d,2s'), hand('3s,3d')]);
    });
    test('findFlushes', () => {
        expect(findFlushes(hand('3s,2h,3h,2s,5h,7h,9h'))).toEqual([hand('9h,7h,5h,3h,2h')]);
        expect(findFlushes(hand('10h,2h,3h,2s,5h,7h,9h'))).toEqual([hand('10h,9h,7h,5h,3h')]);
    });
    test('findStraights', () => {
        expect(findStraights(hand('3s,2h,6h,9s,5h,8h,4d'))).toEqual([hand('2h,3s,4d,5h,6h')]);
        expect(findStraights(hand('3s,2h,6h,9s,5h,7h,4d'))).toEqual([hand('3s,4d,5h,6h,7h'), hand('2h,3s,4d,5h,6h')]);
        expect(findStraights(hand('3s,2h,6h,2s,5h,8h,4d'))).toEqual([hand('2h,3s,4d,5h,6h'), hand('2s,3s,4d,5h,6h')]);
    });
    test('bestHand', () => {
        expect(bestHand(hand('3s,2h,6h,9s,5h,8h,4d'))).toEqual([hand('2h,3s,4d,5h,6h'), 4]);
        expect(bestHand(hand('10h,2h,3h,2s,5h,7h,9h'))).toEqual([hand('10h,9h,7h,5h,3h'), 5]);
        expect(bestHand(hand('2h,10h,3s,2d,2s,4d,3d'))).toEqual([hand('2h,2d,2s,3s,3d'), 6]);
        expect(bestHand(hand('2h,10h,3s,2d,5s,4d,3d'))).toEqual([hand('3s,3d,2h,2d'), 2]);
        expect(bestHand(hand('3h,10h,3s,2d,5s,4d,3d'))).toEqual([hand('3h,3s,3d'), 3]);
        expect(bestHand(hand('3s,2h,6h,9s,5h,8h,10d'))).toEqual([hand('10d'), 0]);
    });
    test('compareHands', () => {
        expect(compareHands(hand('3h,10h,3s,2d,5s,4d,3d'), hand('3s,2h,6h,9s,5h,8h,10d')).aWon).toBe(true);
        expect(compareHands(hand('3s,2h,6h,9s,5h,8h,10d'), hand('3h,10h,3s,2d,5s,4d,3d')).aWon).toBe(false);
    });

    test('probabilities', () => {
        let numStraights = 0;
        let numFlushes = 0;
        let numPairs = 0;
        let numTwoPairs = 0;
        let numThrees = 0;
        let numFullHouses = 0;
        let numNothing = 0;
        let numAvailStraights = 0;
        let numAvailFlushes = 0;
        let numAvailPairs = 0;
        let numAvailTwoPairs = 0;
        let numAvailThrees = 0;
        let numAvailFullHouses = 0;
        let numAvailNothing = 0;
        const runs = 10000;
        const rng = new Rng('pinball');
        for (let i=0; i<runs; i++) {
            const deck = Poker.makeDeck(rng).slice(0, 17+2);
            {
                const play = deck.slice();
                play.shuffle(() => rng.rand());
                const hand = play.slice(0, 7);
                const flushes = findFlushes(hand);
                const straights = findStraights(hand);
                const pairs = findPairs(hand);
                const pair2 = pairs.filter(x => x.length === 2);
                const pair3 = pairs.filter(x => x.length === 3);
                numStraights += straights.length > 0? 1:0;
                numFlushes += flushes.length > 0? 1:0;
                numPairs += pair2.length > 0? 1:0;
                numThrees += pair3.length > 0? 1:0;
                numFullHouses += pair3.length * pair2.length > 0? 1:0;
                numTwoPairs += pair2.length * (pair2.length-1) > 0? 1:0;
                numNothing += !straights.length && !flushes.length && !pairs.length? 1:0;
            }
            {
                const flushes = findFlushes(deck);
                const straights = findStraights(deck);
                const pairs = findPairs(deck);
                const pair2 = pairs.filter(x => x.length === 2);
                const pair3 = pairs.filter(x => x.length === 3);
                numAvailStraights += straights.length > 1? 1:0;
                numAvailFlushes += flushes.length > 1? 1:0;
                numAvailPairs += pair2.length > 0? 1:0;
                numAvailThrees += pair3.length > 0? 1:0;
                numAvailFullHouses += pair3.length * pair2.length > 0? 1:0;
                numAvailTwoPairs += pair2.length * (pair2.length-1) > 0? 1:0;
                numAvailNothing += !straights.length && !flushes.length && !pairs.length? 1:0;
            }
        }
        console.log('Drawn Straights: %i    %f%%', numStraights, (numStraights/runs*100).toFixed(0));
        console.log('Drawn Flushes: %i    %f%%', numFlushes, (numFlushes/runs*100).toFixed(0));
        console.log('Drawn Pairs: %i    %f%%', numPairs, (numPairs/runs*100).toFixed(0));
        console.log('Drawn TwoPairs: %i    %f%%', numTwoPairs, (numTwoPairs/runs*100).toFixed(0));
        console.log('Drawn Threes: %i    %f%%', numThrees, (numThrees/runs*100).toFixed(0));
        console.log('Drawn FullHouses: %i    %f%%', numFullHouses, (numFullHouses/runs*100).toFixed(0));
        console.log('Drawn Nothing: %i    %f%%', numNothing, (numNothing/runs*100).toFixed(0));
        console.log('Avail Straights: %i    %f%%', numAvailStraights, (numAvailStraights/runs*100).toFixed(0));
        console.log('Avail Flushes: %i    %f%%', numAvailFlushes, (numAvailFlushes/runs*100).toFixed(0));
        console.log('Avail Pairs: %i    %f%%', numAvailPairs, (numAvailPairs/runs*100).toFixed(0));
        console.log('Avail TwoPairs: %i    %f%%', numAvailTwoPairs, (numAvailTwoPairs/runs*100).toFixed(0));
        console.log('Avail Threes: %i    %f%%', numAvailThrees, (numAvailThrees/runs*100).toFixed(0));
        console.log('Avail FullHouses: %i    %f%%', numAvailFullHouses, (numAvailFullHouses/runs*100).toFixed(0));
        console.log('Avail Nothing: %i    %f%%', numAvailNothing, (numAvailNothing/runs*100).toFixed(0));
    });

    
    test('close diverter during deal', async () => {
        await testRecording('close gate during deal', 'break');
        expect(machine.cShooterDiverter.val).toBe(true);
        await finishRecording();
        await passTime(10);
        expect(machine.cShooterDiverter.val).toBe(false);
        await passTime(5000);
        expect(machine.cShooterDiverter.val).toBe(true);
        snapshot();
    });

    test('dont eject during cards', async () => {
        await testRecording('upperEjectBlock');
        await passTime(1000);
        expect(machine.cUpperEject.val).toBe(false);
        expect(machine.cUpperEject.lastActualChange).toBe(undefined);
        snapshot();
    });
});