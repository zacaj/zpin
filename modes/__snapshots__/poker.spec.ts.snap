// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`poker close diverter during deal 1`] = `
Object {
  "outputs": Object {
    "centerBank": false,
    "getSkillshot": [Function],
    "iCenter1": "8H",
    "iCenter2": "QS",
    "iCenter3": "QD",
    "iLeft1": "AH",
    "iLeft2": "JD",
    "iLeft3": "QH",
    "iLeft4": "7S",
    "iMini1": "5C",
    "iMini2": "7D",
    "iMini3": "10H",
    "iRight1": "KH",
    "iRight2": "7H",
    "iRight3": "JS",
    "iRight4": "5D",
    "iRight5": "6S",
    "iSS1": Object {
      "fill": [Function],
    },
    "iSS2": Object {
      "fill": [Function],
    },
    "iSS3": Object {
      "fill": [Function],
    },
    "iSS4": Object {
      "fill": [Function],
    },
    "iSS5": Object {
      "fill": [Function],
    },
    "iSS6": Object {
      "fill": [Function],
    },
    "iSS7": Object {
      "fill": [Function],
    },
    "iSpinner": "",
    "iUpper21": "JH",
    "iUpper22": "7C",
    "iUpper31": "10C",
    "iUpper32": "5H",
    "iUpper33": "9H",
    "ignoreSkillsot": Set {},
    "kickerEnable": true,
    "lEjectArrow": Array [],
    "lEjectShowCards": Array [],
    "lEjectStartMode": Array [],
    "lLaneUpper1": Array [
      "Orange",
    ],
    "lLaneUpper2": Array [],
    "lLaneUpper3": Array [],
    "lLaneUpper4": Array [],
    "lLeftArrow": Array [],
    "lLowerRamp": Array [],
    "lMagnaSaveStatus": Array [
      "Green",
    ],
    "lMainTargetArrow": Array [],
    "lMiniReady": Array [
      "Green",
    ],
    "lPopperStatus": Array [
      "Green",
    ],
    "lPower1": Array [
      "Orange",
    ],
    "lPower2": Array [
      "Orange",
    ],
    "lPower3": Array [
      "Orange",
    ],
    "lPower4": Array [
      "Orange",
    ],
    "lRampArrow": Array [],
    "lRampShowCards": Array [],
    "lRampStartMb": Array [],
    "lShooterLaneArrow": Array [],
    "lShooterShowCards": Array [],
    "lShooterStartHand": Array [],
    "lSideShotArrow": Array [],
    "lSideTargetArrow": Array [],
    "lSpinnerArrow": Array [],
    "lUpperLaneArrow": Array [],
    "lUpperTargetArrow": Array [],
    "leftBank": false,
    "leftGate": false,
    "leftMagnet": false,
    "lockPost": false,
    "magnetPost": false,
    "miniBank": false,
    "miniDiverter": false,
    "miniEject": false,
    "miniFlipperEnable": false,
    "outhole": false,
    "popper": false,
    "rampUp": true,
    "realRightBank": false,
    "right1": false,
    "right2": false,
    "right3": false,
    "right4": false,
    "right5": false,
    "rightBank": false,
    "rightGate": false,
    "shooterDiverter": true,
    "troughRelease": false,
    "upper2": false,
    "upper3": false,
    "upperEject": false,
    "upperMagnet": false,
  },
  "state": Object {
    "children": Array [
      Object {
        "children": Array [],
        "name": "DropBank65",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank66",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank67",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank68",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank69",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank70",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "name": "EosPulse72",
            "state": Object {},
          },
        ],
        "name": "MachineOverrides71",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "ClearHoles75",
                "state": Object {},
              },
              Object {
                "children": Array [],
                "name": "Spinner77",
                "state": Object {
                  "comboMult": 1,
                  "rounds": 0,
                  "score": 300,
                },
              },
              Object {
                "children": Array [],
                "name": "LeftOrbit78",
                "state": Object {
                  "comboMult": 1,
                  "rounds": 1,
                  "score": 20000,
                },
              },
              Object {
                "children": Array [],
                "name": "PlayerOverrides79",
                "state": Object {},
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter82",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter83",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter84",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter85",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter86",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter87",
                        "state": Object {},
                      },
                    ],
                    "name": "ResetAnyDropOnComplete81",
                    "state": Object {},
                  },
                  Object {
                    "children": Array [],
                    "name": "Skillshot93",
                    "state": Object {
                      "curAward": 0,
                      "shooterOpen": true,
                    },
                  },
                ],
                "name": "Ball80",
                "state": Object {
                  "miniReady": true,
                },
              },
              Object {
                "children": Array [],
                "name": "Poker92",
                "state": Object {
                  "bet": 100,
                  "closeShooter": false,
                  "dealerCardsUsed": Array [],
                  "dealerHand": Array [
                    Object {
                      "flipped": true,
                      "num": 3,
                      "suit": "d",
                    },
                    Object {
                      "flipped": true,
                      "num": 6,
                      "suit": "h",
                    },
                    null,
                    null,
                    null,
                    null,
                    null,
                  ],
                  "dealerHandDesc": undefined,
                  "newMbs": Array [],
                  "newModes": Array [],
                  "playerCardsUsed": Array [],
                  "playerHand": Array [
                    Object {
                      "num": 3,
                      "suit": "h",
                    },
                    Object {
                      "num": 1,
                      "suit": "d",
                    },
                    null,
                    null,
                    null,
                    null,
                    null,
                  ],
                  "playerHandDesc": undefined,
                  "playerWins": undefined,
                  "pot": 0,
                  "slots": Array [
                    Object {
                      "num": 8,
                      "suit": "h",
                    },
                    Object {
                      "num": 12,
                      "suit": "s",
                    },
                    Object {
                      "num": 12,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "h",
                    },
                    Object {
                      "num": 7,
                      "suit": "h",
                    },
                    Object {
                      "num": 11,
                      "suit": "s",
                    },
                    Object {
                      "num": 5,
                      "suit": "d",
                    },
                    Object {
                      "num": 6,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "h",
                    },
                    Object {
                      "num": 11,
                      "suit": "d",
                    },
                    Object {
                      "num": 12,
                      "suit": "h",
                    },
                    Object {
                      "num": 7,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "h",
                    },
                    Object {
                      "num": 7,
                      "suit": "c",
                    },
                    Object {
                      "num": 10,
                      "suit": "c",
                    },
                    Object {
                      "num": 5,
                      "suit": "h",
                    },
                    Object {
                      "num": 9,
                      "suit": "h",
                    },
                    Object {
                      "num": 5,
                      "suit": "c",
                    },
                    Object {
                      "num": 7,
                      "suit": "d",
                    },
                    Object {
                      "num": 10,
                      "suit": "h",
                    },
                  ],
                  "step": 2,
                },
              },
            ],
            "name": "Player74",
            "state": Object {
              "chips": 4,
              "closeShooter": false,
              "curMbMode": undefined,
              "laneChips": Array [
                true,
                false,
                false,
                false,
              ],
              "mbsQualified": Array [],
              "modesQualified": Array [],
              "noMode": undefined,
              "poker": "Poker92",
              "rampUp": true,
              "score": 31000,
            },
          },
        ],
        "name": "Game73",
        "state": Object {
          "ballNum": 1,
        },
      },
    ],
    "name": "Machine64",
    "state": Object {
      "lastSwitchHit": Switch {
        "_state": true,
        "column": 4,
        "lastChange": 11863,
        "lastClosed": 11863,
        "lastOpened": undefined,
        "minOffTime": 150,
        "minOnTime": 10,
        "name": "right 2",
        "row": 2,
      },
      "lockDown": false,
      "miniDown": false,
    },
  },
}
`;

exports[`poker dont eject during cards 1`] = `
Object {
  "outputs": Object {
    "centerBank": false,
    "getSkillshot": [Function],
    "iCenter1": "",
    "iCenter2": "",
    "iCenter3": "",
    "iLeft1": "",
    "iLeft2": "",
    "iLeft3": "",
    "iLeft4": "",
    "iMini1": "",
    "iMini2": "",
    "iMini3": "",
    "iRight1": "",
    "iRight2": "",
    "iRight3": "",
    "iRight4": "",
    "iRight5": "",
    "iSS1": "",
    "iSS2": "",
    "iSS3": "",
    "iSS4": "",
    "iSS5": "",
    "iSS6": "",
    "iSS7": "",
    "iSpinner": "",
    "iUpper21": "",
    "iUpper22": "",
    "iUpper31": "",
    "iUpper32": "",
    "iUpper33": "",
    "ignoreSkillsot": Set {},
    "kickerEnable": true,
    "lEjectArrow": Array [],
    "lEjectShowCards": Array [
      "White",
    ],
    "lEjectStartMode": Array [],
    "lLaneUpper1": Array [
      "Orange",
    ],
    "lLaneUpper2": Array [],
    "lLaneUpper3": Array [],
    "lLaneUpper4": Array [],
    "lLeftArrow": Array [],
    "lLowerRamp": Array [],
    "lMagnaSaveStatus": Array [
      "Green",
    ],
    "lMainTargetArrow": Array [],
    "lMiniReady": Array [
      "Green",
    ],
    "lPopperStatus": Array [
      "Green",
    ],
    "lPower1": Array [
      "Orange",
    ],
    "lPower2": Array [
      "Orange",
    ],
    "lPower3": Array [
      "Orange",
    ],
    "lPower4": Array [
      "Orange",
    ],
    "lRampArrow": Array [],
    "lRampShowCards": Array [
      "White",
    ],
    "lRampStartMb": Array [],
    "lShooterLaneArrow": Array [],
    "lShooterShowCards": Array [
      "White",
    ],
    "lShooterStartHand": Array [
      Array [
        "Green",
        "fl",
      ],
    ],
    "lSideShotArrow": Array [],
    "lSideTargetArrow": Array [],
    "lSpinnerArrow": Array [],
    "lUpperLaneArrow": Array [],
    "lUpperTargetArrow": Array [],
    "leftBank": false,
    "leftGate": false,
    "leftMagnet": false,
    "lockPost": false,
    "magnetPost": false,
    "miniBank": false,
    "miniDiverter": false,
    "miniEject": false,
    "miniFlipperEnable": false,
    "outhole": false,
    "popper": false,
    "rampUp": false,
    "realRightBank": false,
    "right1": false,
    "right2": false,
    "right3": false,
    "right4": false,
    "right5": false,
    "rightBank": false,
    "rightGate": true,
    "shooterDiverter": true,
    "troughRelease": false,
    "upper2": false,
    "upper3": false,
    "upperEject": false,
    "upperMagnet": false,
  },
  "state": Object {
    "children": Array [
      Object {
        "children": Array [],
        "name": "DropBank104",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank105",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank106",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank107",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank108",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank109",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "name": "EosPulse111",
            "state": Object {},
          },
        ],
        "name": "MachineOverrides110",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "ClearHoles114",
                "state": Object {},
              },
              Object {
                "children": Array [],
                "name": "Poker115",
                "state": Object {
                  "bet": 100,
                  "closeShooter": false,
                  "dealerCardsUsed": Array [
                    Object {
                      "flipped": false,
                      "num": 9,
                      "suit": "h",
                    },
                    Object {
                      "flipped": false,
                      "num": 9,
                      "suit": "d",
                    },
                    Object {
                      "flipped": undefined,
                      "num": 4,
                      "suit": "c",
                    },
                    Object {
                      "flipped": false,
                      "num": 4,
                      "suit": "h",
                    },
                  ],
                  "dealerHand": Array [
                    Object {
                      "flipped": undefined,
                      "num": 3,
                      "suit": "h",
                    },
                    Object {
                      "flipped": undefined,
                      "num": 4,
                      "suit": "c",
                    },
                    Object {
                      "flipped": false,
                      "num": 9,
                      "suit": "h",
                    },
                    Object {
                      "flipped": false,
                      "num": 5,
                      "suit": "s",
                    },
                    Object {
                      "flipped": false,
                      "num": 9,
                      "suit": "d",
                    },
                    Object {
                      "flipped": false,
                      "num": 4,
                      "suit": "h",
                    },
                    Object {
                      "flipped": undefined,
                      "num": 2,
                      "suit": "c",
                    },
                  ],
                  "dealerHandDesc": "Two Pair",
                  "newMbs": Array [],
                  "newModes": Array [],
                  "playerCardsUsed": Array [
                    Object {
                      "num": 13,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "h",
                    },
                  ],
                  "playerHand": Array [
                    Object {
                      "num": 7,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "s",
                    },
                    Object {
                      "num": 13,
                      "suit": "d",
                    },
                    Object {
                      "num": 3,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "h",
                    },
                    Object {
                      "num": 1,
                      "suit": "s",
                    },
                    Object {
                      "num": 9,
                      "suit": "c",
                    },
                  ],
                  "playerHandDesc": "Pair",
                  "playerWins": false,
                  "pot": 1000,
                  "slots": Array [
                    Object {
                      "num": 1,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "c",
                    },
                    Object {
                      "num": 13,
                      "suit": "s",
                    },
                    Object {
                      "num": 8,
                      "suit": "h",
                    },
                    Object {
                      "num": 5,
                      "suit": "h",
                    },
                    Object {
                      "num": 4,
                      "suit": "s",
                    },
                    Object {
                      "num": 5,
                      "suit": "d",
                    },
                    Object {
                      "num": 3,
                      "suit": "s",
                    },
                    Object {
                      "num": 8,
                      "suit": "c",
                    },
                    Object {
                      "num": 10,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "h",
                    },
                    Object {
                      "num": 8,
                      "suit": "s",
                    },
                    Object {
                      "num": 10,
                      "suit": "h",
                    },
                    Object {
                      "num": 12,
                      "suit": "d",
                    },
                    Object {
                      "num": 8,
                      "suit": "d",
                    },
                    Object {
                      "num": 10,
                      "suit": "d",
                    },
                    Object {
                      "num": 12,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "h",
                    },
                    Object {
                      "num": 9,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "c",
                    },
                  ],
                  "step": 8,
                },
              },
              Object {
                "children": Array [],
                "name": "Spinner116",
                "state": Object {
                  "comboMult": 1,
                  "rounds": 0,
                  "score": 10,
                },
              },
              Object {
                "children": Array [],
                "name": "LeftOrbit117",
                "state": Object {
                  "comboMult": 1,
                  "rounds": 1,
                  "score": 20000,
                },
              },
              Object {
                "children": Array [],
                "name": "PlayerOverrides118",
                "state": Object {},
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter121",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter122",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter123",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter124",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter125",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter126",
                        "state": Object {},
                      },
                    ],
                    "name": "ResetAnyDropOnComplete120",
                    "state": Object {},
                  },
                ],
                "name": "Ball119",
                "state": Object {
                  "miniReady": true,
                },
              },
            ],
            "name": "Player113",
            "state": Object {
              "chips": 4,
              "closeShooter": false,
              "curMbMode": undefined,
              "laneChips": Array [
                true,
                false,
                false,
                false,
              ],
              "mbsQualified": Array [],
              "modesQualified": Array [],
              "noMode": undefined,
              "poker": "Poker115",
              "rampUp": true,
              "score": 26000,
            },
          },
        ],
        "name": "Game112",
        "state": Object {
          "ballNum": 1,
        },
      },
    ],
    "name": "Machine103",
    "state": Object {
      "lastSwitchHit": Switch {
        "_state": true,
        "column": 6,
        "lastChange": 16474,
        "lastClosed": 16474,
        "lastOpened": undefined,
        "minOffTime": 25,
        "minOnTime": 25,
        "name": "upper eject",
        "row": 7,
      },
      "lockDown": false,
      "miniDown": false,
    },
  },
}
`;
